internal/transport/http2_client.go

```diff
@@ -1303,29 +1311,20 @@ func (t *http2Client) keepalive() {
 				timer.Reset(t.kp.Time)
 				continue
 			}
-			// Check if keepalive should go dormant.
 			t.mu.Lock()
-			if len(t.activeStreams) < 1 && !t.kp.PermitWithoutStream {
-				// Make awakenKeepalive writable.
-				<-t.awakenKeepalive
-				t.mu.Unlock()
-				select {
-				case <-t.awakenKeepalive:
-					// If the control gets here a ping has been sent
-					// need to reset the timer with keepalive.Timeout.
-				case <-t.ctx.Done():
-					return
-				}
-			} else {
-				t.mu.Unlock()
-				if channelz.IsOn() {
-					atomic.AddInt64(&t.czData.kpCount, 1)
-				}
-				// Send ping.
-				t.controlBuf.put(p)
+			if !t.shouldKeepaliveBeActive() {
+				t.kpDormancyCond.Wait()
+			}
+			t.mu.Unlock()
+
+			if channelz.IsOn() {
+				atomic.AddInt64(&t.czData.kpCount, 1)
 			}
+			// We get here either because we were dormant and a new stream was
+			// created which unblocked the Wait() call, or because the
+			// keepalive timer expired. In both cases, we need to send a ping.
```

Comment:
...or because the transport is shutting down.  Is it harmless to add the ping to the controlbuf in that case?  If not, we can check `t.state` before releasing the lock, above.

Reply:
I checked `controlbuf` code, and looks to me that it shouldn't be a problem. 'put` will return an error if the transport is already closed. But we are anyways ignoring it.